gazelle/generate.go

```diff
@@ -0,0 +1,309 @@
+package gazelle
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/bazelbuild/bazel-gazelle/label"
+	"github.com/bazelbuild/bazel-gazelle/language"
+	"github.com/bazelbuild/bazel-gazelle/rule"
+	"github.com/emirpasic/gods/lists/singlylinkedlist"
+	"github.com/emirpasic/gods/sets/treeset"
+)
+
+const (
+	// The filename (with any of the TS extensions) imported when importing a directory
+	indexFileName = "index"
+)
+
+// GenerateRules extracts build metadata from source files in a directory.
+// GenerateRules is called in each directory where an update is requested
+// in depth-first post-order.
+func (ts *TypeScript) GenerateRules(args language.GenerateArgs) language.GenerateResult {
+	cfgs := args.Config.Exts[languageName].(Configs)
+	cfg := cfgs[args.Rel]
+
+	// When we return empty, we mean that we don't generate anything, but this
+	// still triggers the indexing for all the TypeScript targets in this
+	// package.
+	if !cfg.GenerationEnabled() {
+		return language.GenerateResult{}
+	}
+
+	// If this directory has not been declared as a bazel package it will have been
+	// including in the parent BUILD file.
+	if !isBazelPackage(args.Dir) {
+		return language.GenerateResult{}
+	}
+
+	// Collect all source files
+	sourceFiles, dataFiles, collectErr := collectSourceFiles(cfg, args)
+	if collectErr != nil {
+		log.Printf("ERROR: %v\n", collectErr)
+		return language.GenerateResult{}
+	}
+
+	DEBUG("SOURCE(%q): %s", args.Rel, sourceFiles.Values())
+
+	// Divide src vs test files
+	libSourceFiles := treeset.NewWithStringComparator()
+	testSourceFiles := treeset.NewWithStringComparator()
+
+	for _, f := range sourceFiles.Values() {
+		file := f.(string)
+		if cfg.IsTestFile(file) {
+			testSourceFiles.Add(file)
+		} else if cfg.IsSourceFile(file) {
+			libSourceFiles.Add(file)
+		}
+	}
+
+	// Build the GenerateResult with src and test rules
+	var result language.GenerateResult
+
+	addProjectRule(
+		cfg,
+		args,
+		cfg.RenderLibraryName(filepath.Base(args.Dir)),
+		libSourceFiles,
+		dataFiles,
+		&result,
```

Comment:
This is an OO anti-pattern. You should not modify an argument passed to a function/method.

Reply:
I've kept this for now. The `result` has multiple things in it and is passed into this method multiple times (one for source, one for tests etc).